src/ops/indexing.rs

```diff
@@ -0,0 +1,1799 @@
+            let mut new_shape = vec![1; max_dims + slice_count];
+            new_shape[i] = item.dim(0);
+            *item = Rc::new(item.reshape(&new_shape));
+        }
+    }
+
+    // Do the gather
+    // let indices = new_mlx_vector_array(gather_indices);
+    // SAFETY: indices will be freed at the end of this function. The lifetime of the items in
+    // `gather_indices` is managed by the `gather_indices` vector.
+    let indices = unsafe {
+        let c_vec_array = mlx_sys::mlx_vector_array_new();
+        for item in gather_indices.iter() {
+            mlx_sys::mlx_vector_array_add(c_vec_array, item.c_array);
+        }
+        c_vec_array
+    };
+    let mut slice_sizes = shape.to_vec();
+    (0..operation_len).for_each(|i| slice_sizes[i] = 1);
+
+    let gathered = unsafe {
+        let c_array = mlx_sys::mlx_gather(
+            src.c_array,
+            indices,
+            axes.as_ptr(),
+            axes.len(),
+            slice_sizes.as_ptr(),
+            slice_sizes.len(),
+            stream.as_ptr(),
+        );
+
+        Array::from_ptr(c_array)
+    };
+    let gathered_shape = gathered.shape();
+
+    // Squeeze the dims
+    let output_shape: Vec<i32> = gathered_shape[0..(max_dims + slice_count)]
+        .iter()
+        .chain(gathered_shape[(max_dims + slice_count + operation_len)..].iter())
+        .copied()
+        .collect();
+    let result = gathered.reshape(&output_shape);
+
+    unsafe {
+        mlx_sys::mlx_free(indices as *mut _);
+    }
+
+    (max_dims, result)
+}
+
+#[inline]
+fn get_item_int(src: &Array, index: i32, axis: i32, stream: StreamOrDevice) -> Array {
+    src.take_device(&index.into(), axis, stream)
+}
+
+#[inline]
+fn get_item_array(src: &Array, indices: &Array, axis: i32, stream: StreamOrDevice) -> Array {
+    src.take_device(indices, axis, stream)
+}
+
+#[inline]
+fn get_item_slice(src: &Array, range: RangeIndex, stream: StreamOrDevice) -> Array {
+    let ndim = src.ndim();
+    let mut starts: SmallVec<[i32; 4]> = smallvec![0; ndim];
+    let mut ends: SmallVec<[i32; 4]> = SmallVec::from_slice(src.shape());
+    let mut strides: SmallVec<[i32; 4]> = smallvec![1; ndim];
+
+    let size = ends[0];
+    starts[0] = range.start(size);
+    ends[0] = range.end(size);
+    strides[0] = range.stride();
+
+    src.slice_device(&starts, &ends, &strides, stream)
+}
+
+fn count_non_new_axis_operations(operations: &[ArrayIndexOp]) -> usize {
+    operations
+        .iter()
+        .filter(|op| !matches!(op, ArrayIndexOp::ExpandDims))
+        .count()
+}
+
+fn expand_ellipsis_operations(ndim: usize, operations: &[ArrayIndexOp]) -> Cow<'_, [ArrayIndexOp]> {
+    let ellipsis_count = operations
+        .iter()
+        .filter(|op| matches!(op, ArrayIndexOp::Ellipsis))
+        .count();
+    if ellipsis_count == 0 {
+        return Cow::Borrowed(operations);
+    }
+
+    if ellipsis_count > 1 {
+        panic!("Indexing with multiple ellipsis is not supported");
```

Comment:
Should we be bubbling up this error?

Reply:
I don't think so. Indexing isn't supposed to give an `Result` but just panics if indices are invalid